Could not write to device, result == -1 errno 0 null

Hi, @magnusja 
Libaums version 0.7.3

Hardware devices: 
- usb device 1: SanDisk dual-drive **128Gb** with Fat32, 32/64 Kb cluster size;
- usb device 2: JetFlash **8Gb** with Fat32, 32/4 Kb cluster size;
- android version 9(Samsung Galaxy S10e)/8(Xiaomi Redmi 5 Plus)
### Problem
I try to copy about 1000 mixed size/type files (total size 2 Gb).
### Expected behavior
Files were copied correctly.
### Actual behavior
`java.io.IOException: Could not write to device, result == -1 errno 0 null`
#### Stacktrace of Excpetion
![image](https://user-images.githubusercontent.com/32107577/70522475-3d6c5700-1b52-11ea-865d-d6b9ecfa2157.png)
### Code where problem occurs
UsbFileOutputStream.write(bytes);

Any hints!

Thanks a lot!

Nope, the issue eventually occurred and it looks like the patch didn't help:

```
D/ScsiBlockDevice: inquiry response: ScsiInquiryResponse [peripheralQualifier=0, peripheralDeviceType=0, removableMedia=false, spcVersion=6, responseDataFormat=2]
I/ScsiBlockDevice: Block size: 512
    Last block address: 31735807
E/ScsiBlockDevice: error transferring command, sending bulk only reset
E/ScsiBlockDevice: error transferring command, sending bulk only reset
E/ScsiBlockDevice: error transferring command, sending bulk only reset
E/UsbApiImgWriteService: Could't write image to Depau EtchDrive
    eu.depau.etchdroid.utils.exception.UsbWriteException: Write failed at block 1525760, 781.2 MB written. Error: java.io.IOException: Could not write to device, result == -1 errno 0 null
        at eu.depau.etchdroid.services.UsbApiWriteService.writeInputStream(UsbApiWriteService.kt:76)
        at eu.depau.etchdroid.services.UsbApiWriteService.writeImage(UsbApiWriteService.kt:107)
        at eu.depau.etchdroid.services.UsbWriteService.onHandleIntent(UsbWriteService.kt:36)
        at android.app.IntentService$ServiceHandler.handleMessage(IntentService.java:78)
        at android.os.Handler.dispatchMessage(Handler.java:107)
        at android.os.Looper.loop(Looper.java:214)
        at android.os.HandlerThread.run(HandlerThread.java:67)
     Caused by: java.io.IOException: Could not write to device, result == -1 errno 0 null
        at com.github.mjdev.libaums.usb.JellyBeanMr2Communication.bulkOutTransfer(JellyBeanMr2Communication.kt:35)
        at com.github.mjdev.libaums.driver.scsi.ScsiBlockDevice.transferOneCommand(ScsiBlockDevice.kt:186)
        at com.github.mjdev.libaums.driver.scsi.ScsiBlockDevice.transferCommand(ScsiBlockDevice.kt:140)
        at com.github.mjdev.libaums.driver.scsi.ScsiBlockDevice.write(ScsiBlockDevice.kt:250)
        at eu.depau.etchdroid.services.UsbApiWriteService.writeInputStream(UsbApiWriteService.kt:74)
        at eu.depau.etchdroid.services.UsbApiWriteService.writeImage(UsbApiWriteService.kt:107) 
        at eu.depau.etchdroid.services.UsbWriteService.onHandleIntent(UsbWriteService.kt:36) 
        at android.app.IntentService$ServiceHandler.handleMessage(IntentService.java:78) 
        at android.os.Handler.dispatchMessage(Handler.java:107) 
        at android.os.Looper.loop(Looper.java:214) 
        at android.os.HandlerThread.run(HandlerThread.java:67) 
D/UsbMassStorageDevice: close device
D/UsbDeviceConnectionJNI: close
```

> thanks a lot for testing. Well that is too bad :/ I will implement the first option, we can still think about the second one. We can also only move the reset then to native code. What essentially is happening is this: https://android.googlesource.com/platform/system/core/+/master/libusbhost/usbhost.c?autodive=0#682 This can be easily implement keeping the whole Kotlin (as of now :D) code!

Yeah, an IOCTL should be doable from Java/Kotlin.

Anyway, nope, it doesn't help:

```
I/UsbMassStorageDevice: MAX LUN 0
D/ScsiBlockDevice: inquiry response: ScsiInquiryResponse [peripheralQualifier=0, peripheralDeviceType=0, removableMedia=false, spcVersion=6, responseDataFormat=2]
I/ScsiBlockDevice: Block size: 512
    Last block address: 31735807
W/depau.etchdroi: JNI critical lock held for 65.871ms on Thread[5,tid=8545,Runnable,Thread*=0x6fa76c0c00,peer=0x12fc0040,"IntentService[UsbApiImgWriteService]"]
E/ScsiBlockDevice: error transferring command, sending bulk only reset
W/AndroidUsbCommunication: sending bulk only mass storage request
W/AndroidUsbCommunication: sending clear feature halt
W/AndroidUsbCommunication: sending clear feature halt
E/ScsiBlockDevice: error transferring command, sending bulk only reset
W/AndroidUsbCommunication: sending bulk only mass storage request
W/AndroidUsbCommunication: sending clear feature halt
W/AndroidUsbCommunication: sending clear feature halt
E/ScsiBlockDevice: error transferring command, sending bulk only reset
W/AndroidUsbCommunication: sending bulk only mass storage request
W/AndroidUsbCommunication: sending clear feature halt
W/AndroidUsbCommunication: sending clear feature halt
E/UsbApiImgWriteService: Could't write image to Depau EtchDrive
    eu.depau.etchdroid.utils.exception.UsbWriteException: Write failed at block 0, 0.0 B written. Error: java.io.IOException: Could not write to device, result == -1 errno 0 null
        at eu.depau.etchdroid.services.UsbApiWriteService.writeInputStream(UsbApiWriteService.kt:76)
        at eu.depau.etchdroid.services.UsbApiWriteService.writeImage(UsbApiWriteService.kt:107)
        at eu.depau.etchdroid.services.UsbWriteService.onHandleIntent(UsbWriteService.kt:36)
        at android.app.IntentService$ServiceHandler.handleMessage(IntentService.java:78)
        at android.os.Handler.dispatchMessage(Handler.java:107)
        at android.os.Looper.loop(Looper.java:214)
        at android.os.HandlerThread.run(HandlerThread.java:67)
     Caused by: java.io.IOException: Could not write to device, result == -1 errno 0 null
        at com.github.mjdev.libaums.usb.JellyBeanMr2Communication.bulkOutTransfer(JellyBeanMr2Communication.kt:35)
        at com.github.mjdev.libaums.driver.scsi.ScsiBlockDevice.transferOneCommand(ScsiBlockDevice.kt:191)
        at com.github.mjdev.libaums.driver.scsi.ScsiBlockDevice.transferCommand(ScsiBlockDevice.kt:142)
        at com.github.mjdev.libaums.driver.scsi.ScsiBlockDevice.write(ScsiBlockDevice.kt:255)
        at eu.depau.etchdroid.services.UsbApiWriteService.writeInputStream(UsbApiWriteService.kt:74)
        at eu.depau.etchdroid.services.UsbApiWriteService.writeImage(UsbApiWriteService.kt:107) 
        at eu.depau.etchdroid.services.UsbWriteService.onHandleIntent(UsbWriteService.kt:36) 
        at android.app.IntentService$ServiceHandler.handleMessage(IntentService.java:78) 
        at android.os.Handler.dispatchMessage(Handler.java:107) 
        at android.os.Looper.loop(Looper.java:214) 
        at android.os.HandlerThread.run(HandlerThread.java:67) 
D/UsbMassStorageDevice: close device
D/UsbDeviceConnectionJNI: close
```

Nope :(

```
D/UsbMassStorageDevice: setup device
I/UsbMassStorageDevice: MAX LUN 0
D/ScsiBlockDevice: inquiry response: ScsiInquiryResponse [peripheralQualifier=0, peripheralDeviceType=0, removableMedia=false, spcVersion=6, responseDataFormat=2]
I/ScsiBlockDevice: Block size: 512
    Last block address: 31735807
E/ScsiBlockDevice: error transferring command, sending bulk only reset
W/AndroidUsbCommunication: sending bulk only mass storage request
W/AndroidUsbCommunication: sending clear feature halt
W/AndroidUsbCommunication: sending clear feature halt
W/AndroidUsbCommunication: Native reset
D/UsbDocumentProvider: detachDevice() UsbDevice[mName=/dev/bus/usb/001/002,mVendorId=3544,mProductId=12801,mClass=0,mSubclass=0,mProtocol=0,mManufacturerName=Depau,mProductName=EtchDrive,mVersion=11.00,mSerialNumberReader=android.hardware.usb.IUsbSerialReader$Stub$Proxy@a11f61b,mConfigurations=[
    UsbConfiguration[mId=1,mName=null,mAttributes=128,mMaxPower=50,mInterfaces=[
    UsbInterface[mId=0,mAlternateSetting=0,mName=null,mClass=8,mSubclass=6,mProtocol=80,mEndpoints=[
    UsbEndpoint[mAddress=1,mAttributes=2,mMaxPacketSize=512,mInterval=0]
    UsbEndpoint[mAddress=130,mAttributes=2,mMaxPacketSize=512,mInterval=0]]]]
E/UsbApiImgWriteService: Could't write image to Depau EtchDrive
    eu.depau.etchdroid.utils.exception.UsbWriteException: Write failed at block 0, 0.0 B written. Error: java.io.IOException: native reset unsuccessful!
        at eu.depau.etchdroid.services.UsbApiWriteService.writeInputStream(UsbApiWriteService.kt:76)
        at eu.depau.etchdroid.services.UsbApiWriteService.writeImage(UsbApiWriteService.kt:107)
        at eu.depau.etchdroid.services.UsbWriteService.onHandleIntent(UsbWriteService.kt:36)
        at android.app.IntentService$ServiceHandler.handleMessage(IntentService.java:78)
        at android.os.Handler.dispatchMessage(Handler.java:107)
        at android.os.Looper.loop(Looper.java:214)
        at android.os.HandlerThread.run(HandlerThread.java:67)
     Caused by: java.io.IOException: native reset unsuccessful!
        at com.github.mjdev.libaums.usb.AndroidUsbCommunication.resetRecovery(AndroidUsbCommunication.kt:41)
        at com.github.mjdev.libaums.driver.scsi.ScsiBlockDevice.transferCommand(ScsiBlockDevice.kt:149)
        at com.github.mjdev.libaums.driver.scsi.ScsiBlockDevice.write(ScsiBlockDevice.kt:255)
        at eu.depau.etchdroid.services.UsbApiWriteService.writeInputStream(UsbApiWriteService.kt:74)
        at eu.depau.etchdroid.services.UsbApiWriteService.writeImage(UsbApiWriteService.kt:107) 
        at eu.depau.etchdroid.services.UsbWriteService.onHandleIntent(UsbWriteService.kt:36) 
        at android.app.IntentService$ServiceHandler.handleMessage(IntentService.java:78) 
        at android.os.Handler.dispatchMessage(Handler.java:107) 
        at android.os.Looper.loop(Looper.java:214) 
        at android.os.HandlerThread.run(HandlerThread.java:67) 
D/UsbMassStorageDevice: close device
E/UsbMassStorageDevice: could not release interface!
D/UsbDeviceConnectionJNI: close
```

Also after the "native reset" the USB does not show up any more when querying for USB devices.

I'm noticing that, for whatever reason, I'm getting almost always successfully writes while back then I would only be able to write ~500MB before it would show the issue.

Anyway I was tired of not seeing errors so I built the utility for Android and used it to reset the USB drive while the app was writing, which triggered the new code of course.

I'm testing with a bunch of commits on top of your changes: https://github.com/Depau/libaums/tree/feature/bulk-only-reset

Kernel log with comments:

```
# Manually performed reset
[ 2483.339314] usb 1-1: reset high-speed USB device number 3 using xhci-hcd
[ 2483.463313] xhci-hcd xhci-hcd.0.auto: xHCI xhci_drop_endpoint called with disabled ep 0000000000000000
[ 2483.463362] xhci-hcd xhci-hcd.0.auto: xHCI xhci_drop_endpoint called with disabled ep 0000000000000000
[ 2483.480856] xhci-hcd xhci-hcd.0.auto: hcd->state:1
[ 2483.482825] usb-storage 1-1:1.0: USB Mass Storage device detected
[ 2483.487379] scsi host11: usb-storage 1-1:1.0
# Resets performed by the app
[ 2483.611162] usb 1-1: reset high-speed USB device number 3 using xhci-hcd
[ 2483.740874] xhci-hcd xhci-hcd.0.auto: xHCI xhci_drop_endpoint called with disabled ep 0000000000000000
[ 2483.740930] xhci-hcd xhci-hcd.0.auto: xHCI xhci_drop_endpoint called with disabled ep 0000000000000000
[ 2483.754864] xhci-hcd xhci-hcd.0.auto: hcd->state:1
[ 2484.982369] scsi 11:0:0:0: Direct-Access     Depau    EtchDrive        1100 PQ: 0 ANSI: 6
[ 2485.001798] sdg: sdg1 sdg2
[ 2487.262678] usb 1-1: usbfs: process 15663 (IntentService[U) did not claim interface 0 before use
[ 2487.372826] usb 1-1: reset high-speed USB device number 3 using xhci-hcd
[ 2487.500100] xhci-hcd xhci-hcd.0.auto: xHCI xhci_drop_endpoint called with disabled ep 0000000000000000
[ 2487.500169] xhci-hcd xhci-hcd.0.auto: xHCI xhci_drop_endpoint called with disabled ep 0000000000000000
[ 2487.508317] xhci-hcd xhci-hcd.0.auto: hcd->state:1
[ 2491.015012] usb 1-1: usbfs: process 15663 (IntentService[U) did not claim interface 0 before use
[ 2491.126074] usb 1-1: reset high-speed USB device number 3 using xhci-hcd
[ 2491.252629] xhci-hcd xhci-hcd.0.auto: xHCI xhci_drop_endpoint called with disabled ep 0000000000000000
[ 2491.252682] xhci-hcd xhci-hcd.0.auto: xHCI xhci_drop_endpoint called with disabled ep 0000000000000000
[ 2491.269768] xhci-hcd xhci-hcd.0.auto: hcd->state:1
[ 2494.777149] usb 1-1: usbfs: process 15663 (IntentService[U) did not claim interface 0 before use
```

All ioctls succeeded so it didn't act like the drive was reinserted:

```
E/ScsiBlockDevice: error transferring command, sending bulk only reset
W/AndroidUsbCommunication: Native reset
W/AndroidUsbCommunication: ioctl returned true; errno 0 null; ignoring
E/ScsiBlockDevice: error transferring command, sending bulk only reset
W/AndroidUsbCommunication: Native reset
W/AndroidUsbCommunication: ioctl returned true; errno 0 null; ignoring
E/ScsiBlockDevice: error transferring command, sending bulk only reset
W/AndroidUsbCommunication: Native reset
W/AndroidUsbCommunication: ioctl returned true; errno 0 null; ignoring
E/UsbApiImgWriteService: Could't write image to Depau EtchDrive
    eu.depau.etchdroid.utils.exception.UsbWriteException: Write failed at block 161792, 82.8 MB written. Error: java.io.IOException: Could not write to device, result == -1 errno 0 null
        at eu.depau.etchdroid.services.UsbApiWriteService.writeInputStream(UsbApiWriteService.kt:76)
        at eu.depau.etchdroid.services.UsbApiWriteService.writeImage(UsbApiWriteService.kt:107)
        at eu.depau.etchdroid.services.UsbWriteService.onHandleIntent(UsbWriteService.kt:36)
        at android.app.IntentService$ServiceHandler.handleMessage(IntentService.java:76)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loop(Looper.java:193)
        at android.os.HandlerThread.run(HandlerThread.java:65)
     Caused by: java.io.IOException: Could not write to device, result == -1 errno 0 null
        at com.github.mjdev.libaums.usb.JellyBeanMr2Communication.bulkOutTransfer(JellyBeanMr2Communication.kt:35)
        at com.github.mjdev.libaums.driver.scsi.ScsiBlockDevice.transferOneCommand(ScsiBlockDevice.kt:170)
        at com.github.mjdev.libaums.driver.scsi.ScsiBlockDevice.transferCommand(ScsiBlockDevice.kt:142)
        at com.github.mjdev.libaums.driver.scsi.ScsiBlockDevice.write(ScsiBlockDevice.kt:255)
        at eu.depau.etchdroid.services.UsbApiWriteService.writeInputStream(UsbApiWriteService.kt:74)
        at eu.depau.etchdroid.services.UsbApiWriteService.writeImage(UsbApiWriteService.kt:107) 
        at eu.depau.etchdroid.services.UsbWriteService.onHandleIntent(UsbWriteService.kt:36) 
        at android.app.IntentService$ServiceHandler.handleMessage(IntentService.java:76) 
        at android.os.Handler.dispatchMessage(Handler.java:106) 
        at android.os.Looper.loop(Looper.java:193) 
        at android.os.HandlerThread.run(HandlerThread.java:65) 
D/UsbMassStorageDevice: close device
E/UsbMassStorageDevice: could not release interface!
D/UsbDeviceConnectionJNI: close
D/UsbApiImgWriteService: Written 0 bytes to Depau EtchDrive using API
W/System: A resource failed to call close. 
```

I'll try moving the opening code inside the communication class, so the USB connection can eventually be reopened after reset.

--- a/libaums/CMakeLists.txt
+++ b/libaums/CMakeLists.txt
@@ -1,9 +1,9 @@
 cmake_minimum_required(VERSION 3.4.1)
 
 add_library( errno-lib
-
-             # Sets the library as a shared library.
              SHARED
-
-             # Provides a relative path to your source file(s).
              src/c/errno.c )
+
+add_library( usb-lib
+             SHARED
+             src/c/usb.c )

--- a/libaums/src/c/usb.c
+++ b/libaums/src/c/usb.c
@@ -7,3 +7,9 @@ Java_com_github_mjdev_libaums_usb_AndroidUsbCommunication_resetUsbDeviceNative(J
     int ret = ioctl(fd, USBDEVFS_RESET);
     return (ret == 0) ? JNI_TRUE : JNI_FALSE;
 }
+
+JNIEXPORT jboolean JNICALL
+Java_com_github_mjdev_libaums_usb_AndroidUsbCommunication_clearHaltNative(JNIEnv *env, jobject thiz, jint fd, jint endpoint) {
+    int ret = ioctl(fd, USBDEVFS_CLEAR_HALT, &endpoint);
+    return (ret == 0) ? JNI_TRUE : JNI_FALSE;
+}

--- a/libaums/src/c/usb.c
+++ b/libaums/src/c/usb.c
@@ -1,0 +1,9 @@
+#include <jni.h>
+#include <sys/ioctl.h>
+#include <linux/usbdevice_fs.h>
+
+JNIEXPORT jboolean JNICALL
+Java_com_github_mjdev_libaums_usb_AndroidUsbCommunication_resetUsbDeviceNative(JNIEnv *env, jobject thiz, jint fd) {
+    int ret = ioctl(fd, USBDEVFS_RESET);
+    return (ret == 0) ? JNI_TRUE : JNI_FALSE;
+}

--- a/libaums/src/main/java/com/github/mjdev/libaums/UsbMassStorageDevice.kt
+++ b/libaums/src/main/java/com/github/mjdev/libaums/UsbMassStorageDevice.kt
@@ -76,12 +76,12 @@ private constructor(private val usbManager: UsbManager,
                     val usbDevice: UsbDevice,
                     private val usbInterface: UsbInterface, private val inEndpoint: UsbEndpoint, private val outEndpoint: UsbEndpoint) {
 
-    private lateinit var deviceConnection: UsbDeviceConnection
-
     lateinit var partitions: List<Partition>
 
     // TODO this is never used, should we only allow one init() call?
     private var inited = false
+    
+    private lateinit var usbCommunication: UsbCommunication
 
     /**
      * Initializes the mass storage device and determines different things like
@@ -122,23 +122,16 @@ private constructor(private val usbManager: UsbManager,
      */
     @Throws(IOException::class)
     private fun setupDevice() {
-        Log.d(TAG, "setup device")
-        deviceConnection = usbManager.openDevice(usbDevice)
-                ?: throw IOException("deviceConnection is null!")
-
-        val claim = deviceConnection.claimInterface(usbInterface, true)
-        if (!claim) {
-            throw IOException("could not claim interface!")
-        }
-
-        val communication = UsbCommunicationFactory.createUsbCommunication(deviceConnection, usbInterface, outEndpoint, inEndpoint)
+        usbCommunication = UsbCommunicationFactory
+                .createUsbCommunication(usbManager, usbDevice, usbInterface, outEndpoint, inEndpoint)
         val maxLun = ByteArray(1)
-        deviceConnection.controlTransfer(161, 254, 0, usbInterface.id, maxLun, 1, 5000)
+        usbCommunication.controlTransfer(161, 254, 0, usbInterface.id, maxLun, 1)
+
         Log.i(TAG, "MAX LUN " + maxLun[0].toInt())
 
         this.partitions = (0..maxLun[0])
                 .map { lun ->
-                    BlockDeviceDriverFactory.createBlockDevice(communication, lun = lun.toByte())
+                    BlockDeviceDriverFactory.createBlockDevice(usbCommunication, lun = lun.toByte())
                 }
                 .mapNotNull { blockDevice ->
                     try {
@@ -180,16 +173,7 @@ private constructor(private val usbManager: UsbManager,
      * or write from or to the partitions returned by [.getPartitions].
      */
     fun close() {
-        Log.d(TAG, "close device")
-
-        if (!::deviceConnection.isInitialized)
-            return
-
-        val release = deviceConnection.releaseInterface(usbInterface)
-        if (!release) {
-            Log.e(TAG, "could not release interface!")
-        }
-        deviceConnection.close()
+        usbCommunication.close()
         inited = false
     }
 

--- a/libaums/src/main/java/com/github/mjdev/libaums/UsbMassStorageDevice.kt
+++ b/libaums/src/main/java/com/github/mjdev/libaums/UsbMassStorageDevice.kt
@@ -131,7 +131,7 @@ private constructor(private val usbManager: UsbManager,
             throw IOException("could not claim interface!")
         }
 
-        val communication = UsbCommunicationFactory.createUsbCommunication(deviceConnection, outEndpoint, inEndpoint)
+        val communication = UsbCommunicationFactory.createUsbCommunication(deviceConnection, usbInterface, outEndpoint, inEndpoint)
         val maxLun = ByteArray(1)
         deviceConnection.controlTransfer(161, 254, 0, usbInterface.id, maxLun, 1, 5000)
         Log.i(TAG, "MAX LUN " + maxLun[0].toInt())

--- a/libaums/src/main/java/com/github/mjdev/libaums/driver/scsi/ScsiBlockDevice.kt
+++ b/libaums/src/main/java/com/github/mjdev/libaums/driver/scsi/ScsiBlockDevice.kt
@@ -148,6 +148,8 @@ class ScsiBlockDevice(private val usbCommunication: UsbCommunication, private va
                 Thread.sleep(1500);
             }
         }
+
+        throw IllegalStateException("This should never happen..")
     }
 
 

--- a/libaums/src/main/java/com/github/mjdev/libaums/driver/scsi/ScsiBlockDevice.kt
+++ b/libaums/src/main/java/com/github/mjdev/libaums/driver/scsi/ScsiBlockDevice.kt
@@ -142,14 +142,30 @@ class ScsiBlockDevice(private val usbCommunication: UsbCommunication, private va
             try {
                 return transferOneCommand(command, inBuffer)
             } catch(e: IOException) {
-                if (i == 3) {
-                    throw e
+                Log.e(TAG, "Error transferring command; errno ${ErrNo.errno} ${ErrNo.errstr}")
+
+                when (i) {
+                    0 -> {
+                        Log.d(TAG, "Trying to clear halt on both endpoints")
+                        usbCommunication.clearFeatureHalt(usbCommunication.inEndpoint)
+                        usbCommunication.clearFeatureHalt(usbCommunication.outEndpoint)
+                    }
+                    1 -> {
+                        Log.d(TAG, "Trying to reset the device")
+                        usbCommunication.resetRecovery()
+                    }
+                    2 -> {
+                        Log.d(TAG, "Trying to clear halt on both endpoints again")
+                        usbCommunication.clearFeatureHalt(usbCommunication.inEndpoint)
+                        usbCommunication.clearFeatureHalt(usbCommunication.outEndpoint)
+                    }
+                    else -> {
+                        Log.d(TAG, "Giving up")
+                        throw e
+                    }
                 }
 
-                Log.e(TAG, "Error transferring command; errno ${ErrNo.errno} ${ErrNo.errstr}")
-                Log.e(TAG, "Performing recovery reset procedure")
-                usbCommunication.resetRecovery()
-                Thread.sleep(1500);
+                Thread.sleep(1000)
             }
         }
 

--- a/libaums/src/main/java/com/github/mjdev/libaums/driver/scsi/ScsiBlockDevice.kt
+++ b/libaums/src/main/java/com/github/mjdev/libaums/driver/scsi/ScsiBlockDevice.kt
@@ -18,6 +18,7 @@
 package com.github.mjdev.libaums.driver.scsi
 
 import android.util.Log
+import com.github.mjdev.libaums.ErrNo
 import com.github.mjdev.libaums.driver.BlockDeviceDriver
 import com.github.mjdev.libaums.driver.scsi.commands.*
 import com.github.mjdev.libaums.driver.scsi.commands.CommandBlockWrapper.Direction
@@ -145,13 +146,14 @@ class ScsiBlockDevice(private val usbCommunication: UsbCommunication, private va
                     throw e
                 }
 
-                Log.e(TAG, "error transferring command, sending bulk only reset")
+                Log.e(TAG, "Error transferring command; errno ${ErrNo.errno} ${ErrNo.errstr}")
+                Log.e(TAG, "Performing recovery reset procedure")
                 usbCommunication.resetRecovery()
                 Thread.sleep(1500);
             }
         }
 
-        throw IllegalStateException("This should never happen.. ")
+        throw IllegalStateException("This should never happen.")
     }
 
 

--- a/libaums/src/main/java/com/github/mjdev/libaums/driver/scsi/ScsiBlockDevice.kt
+++ b/libaums/src/main/java/com/github/mjdev/libaums/driver/scsi/ScsiBlockDevice.kt
@@ -147,6 +147,8 @@ class ScsiBlockDevice(private val usbCommunication: UsbCommunication, private va
                 // Try alternately to clear halt and reset device until something happens
                 when {
                     i % 2 == 0 -> {
+                        Log.d(TAG, "Reset bulk-only mass storage")
+                        usbCommunication.bulkOnlyMassStorageReset()
                         Log.d(TAG, "Trying to clear halt on both endpoints")
                         usbCommunication.clearFeatureHalt(usbCommunication.inEndpoint)
                         usbCommunication.clearFeatureHalt(usbCommunication.outEndpoint)

--- a/libaums/src/main/java/com/github/mjdev/libaums/driver/scsi/ScsiBlockDevice.kt
+++ b/libaums/src/main/java/com/github/mjdev/libaums/driver/scsi/ScsiBlockDevice.kt
@@ -48,6 +48,8 @@ class ScsiBlockDevice(private val usbCommunication: UsbCommunication, private va
     private val readCommand = ScsiRead10(lun=lun)
     private val csw = CommandStatusWrapper()
 
+    private var cbwTagCounter = 0
+
     /**
      * The size of the block device, in blocks of [blockSize] bytes,
      *
@@ -149,7 +151,7 @@ class ScsiBlockDevice(private val usbCommunication: UsbCommunication, private va
             }
         }
 
-        throw IllegalStateException("This should never happen..")
+        throw IllegalStateException("This should never happen.. ")
     }
 
 
@@ -157,6 +159,9 @@ class ScsiBlockDevice(private val usbCommunication: UsbCommunication, private va
     private fun transferOneCommand(command: CommandBlockWrapper, inBuffer: ByteBuffer): Boolean {
         val outArray = outBuffer.array()
         Arrays.fill(outArray, 0.toByte())
+
+        command.dCbwTag = cbwTagCounter
+        cbwTagCounter++
 
         outBuffer.clear()
         command.serialize(outBuffer)

--- a/libaums/src/main/java/com/github/mjdev/libaums/driver/scsi/ScsiBlockDevice.kt
+++ b/libaums/src/main/java/com/github/mjdev/libaums/driver/scsi/ScsiBlockDevice.kt
@@ -138,34 +138,30 @@ class ScsiBlockDevice(private val usbCommunication: UsbCommunication, private va
      */
     @Throws(IOException::class)
     private fun transferCommand(command: CommandBlockWrapper, inBuffer: ByteBuffer): Boolean {
-        for(i in 0..3) {
+        for(i in 0..MAX_RECOVERY_ATTEMPTS) {
             try {
                 return transferOneCommand(command, inBuffer)
             } catch(e: IOException) {
                 Log.e(TAG, "Error transferring command; errno ${ErrNo.errno} ${ErrNo.errstr}")
 
-                when (i) {
-                    0 -> {
+                // Try alternately to clear halt and reset device until something happens
+                when {
+                    i % 2 == 0 -> {
                         Log.d(TAG, "Trying to clear halt on both endpoints")
                         usbCommunication.clearFeatureHalt(usbCommunication.inEndpoint)
                         usbCommunication.clearFeatureHalt(usbCommunication.outEndpoint)
                     }
-                    1 -> {
+                    i % 2 == 1 -> {
                         Log.d(TAG, "Trying to reset the device")
                         usbCommunication.resetRecovery()
                     }
-                    2 -> {
-                        Log.d(TAG, "Trying to clear halt on both endpoints again")
-                        usbCommunication.clearFeatureHalt(usbCommunication.inEndpoint)
-                        usbCommunication.clearFeatureHalt(usbCommunication.outEndpoint)
-                    }
-                    else -> {
+                    i == MAX_RECOVERY_ATTEMPTS -> {
                         Log.d(TAG, "Giving up")
                         throw e
                     }
                 }
 
-                Thread.sleep(1000)
+                Thread.sleep(500)
             }
         }
 
@@ -277,7 +273,7 @@ class ScsiBlockDevice(private val usbCommunication: UsbCommunication, private va
     }
 
     companion object {
-
+        private const val MAX_RECOVERY_ATTEMPTS = 20
         private val TAG = ScsiBlockDevice::class.java.simpleName
     }
 }

--- a/libaums/src/main/java/com/github/mjdev/libaums/driver/scsi/ScsiBlockDevice.kt
+++ b/libaums/src/main/java/com/github/mjdev/libaums/driver/scsi/ScsiBlockDevice.kt
@@ -135,6 +135,24 @@ class ScsiBlockDevice(private val usbCommunication: UsbCommunication, private va
      */
     @Throws(IOException::class)
     private fun transferCommand(command: CommandBlockWrapper, inBuffer: ByteBuffer): Boolean {
+        for(i in 0..3) {
+            try {
+                return transferOneCommand(command, inBuffer)
+            } catch(e: IOException) {
+                if (i == 3) {
+                    throw e
+                }
+
+                Log.e(TAG, "error transferring command, sending bulk only reset")
+                usbCommunication.resetRecovery()
+                Thread.sleep(1500);
+            }
+        }
+    }
+
+
+    @Throws(IOException::class)
+    private fun transferOneCommand(command: CommandBlockWrapper, inBuffer: ByteBuffer): Boolean {
         val outArray = outBuffer.array()
         Arrays.fill(outArray, 0.toByte())
 

--- a/libaums/src/main/java/com/github/mjdev/libaums/usb/AndroidUsbCommunication.kt
+++ b/libaums/src/main/java/com/github/mjdev/libaums/usb/AndroidUsbCommunication.kt
@@ -59,21 +59,22 @@ internal abstract class AndroidUsbCommunication(
 
     private fun nativeReset() {
         Log.d(TAG, "Performing native reset")
-        val result = resetUsbDeviceNative(deviceConnection!!.fileDescriptor)
 
-        Log.d(TAG, "ioctl okay: $result; errno ${ErrNo.errno} ${ErrNo.errstr}")
-        if (!result)
-            Log.w(TAG, "ioctl failed! USB device will likely require new discovery and permissions")
+        if (!deviceConnection!!.releaseInterface(usbInterface)) {
+            Log.w(TAG, "Failed to release interface, errno: ${ErrNo.errno} ${ErrNo.errstr}")
+        }
 
-        Thread.sleep(1000)
+        if (!resetUsbDeviceNative(deviceConnection!!.fileDescriptor)) {
+            Log.w(TAG, "ioctl failed! errno ${ErrNo.errno} ${ErrNo.errstr}")
+            Log.w(TAG, "USB device will likely require new discovery and permissions")
+        }
 
-        Log.d(TAG, "Closing and reopening USB device")
-        closeUsbConnection()
-        initUsbConnection()
+        if (!deviceConnection!!.claimInterface(usbInterface, true)) {
+            throw IOException("Could not claim interface, errno: ${ErrNo.errno} ${ErrNo.errstr}")
+        }
     }
 
     private fun softReset() {
-        Log.w(TAG, "Native helper library not available, can't perform hard device reset")
         bulkOnlyMassStorageReset()
         Thread.sleep(2000)
         clearFeatureHalt(inEndpoint)

--- a/libaums/src/main/java/com/github/mjdev/libaums/usb/AndroidUsbCommunication.kt
+++ b/libaums/src/main/java/com/github/mjdev/libaums/usb/AndroidUsbCommunication.kt
@@ -1,0 +1,40 @@
+package com.github.mjdev.libaums.usb
+
+import android.hardware.usb.UsbDeviceConnection
+import android.hardware.usb.UsbEndpoint
+import android.hardware.usb.UsbInterface
+import com.github.mjdev.libaums.usb.UsbCommunication.Companion.TRANSFER_TIMEOUT
+import java.io.IOException
+
+
+internal abstract class AndroidUsbCommunication(private val deviceConnection: UsbDeviceConnection, private val usbInterface: UsbInterface, private val outEndpoint: UsbEndpoint, private val inEndpoint: UsbEndpoint) : UsbCommunication {
+    override fun controlTransfer(requestType: Int, request: Int, value: Int, index: Int, buffer: ByteArray, length: Int): Int {
+        return deviceConnection.controlTransfer(requestType, request, value, index, buffer, length, TRANSFER_TIMEOUT)
+    }
+
+    override fun resetRecovery() {
+        bulkOnlyMassStorageReset();
+        clearFeatureHalt(inEndpoint);
+    }
+
+    override fun bulkOnlyMassStorageReset() {
+        val bArr = ByteArray(2)
+        // REQUEST_BULK_ONLY_MASS_STORAGE_RESET = 255
+        // REQUEST_TYPE_BULK_ONLY_MASS_STORAGE_RESET = 33
+        val transferred: Int = controlTransfer(33, 255, 0, usbInterface.id, bArr, 0)
+        if (transferred == -1) {
+            throw IOException("bulk only mass storage reset failed!")
+        }
+    }
+
+    override fun clearFeatureHalt(endpoint: UsbEndpoint) {
+        val bArr = ByteArray(2)
+        val address: Int = endpoint.address
+        // REQUEST_CLEAR_FEATURE = 1
+        // REQUEST_TYPE_CLEAR_FEATURE = 2
+        val transferred: Int = controlTransfer(2, 1, 0, address, bArr, 0)
+        if (transferred == -1) {
+            throw IOException("bulk only mass storage reset failed!")
+        }
+    }
+}

--- a/libaums/src/main/java/com/github/mjdev/libaums/usb/AndroidUsbCommunication.kt
+++ b/libaums/src/main/java/com/github/mjdev/libaums/usb/AndroidUsbCommunication.kt
@@ -11,8 +11,8 @@ internal abstract class AndroidUsbCommunication(
         private val usbManager: UsbManager,
         private val usbDevice: UsbDevice,
         private val usbInterface: UsbInterface,
-        val outEndpoint: UsbEndpoint,
-        val inEndpoint: UsbEndpoint
+        override val outEndpoint: UsbEndpoint,
+        override val inEndpoint: UsbEndpoint
 ) : UsbCommunication {
 
     private var isNativeInited: Boolean = false

--- a/libaums/src/main/java/com/github/mjdev/libaums/usb/AndroidUsbCommunication.kt
+++ b/libaums/src/main/java/com/github/mjdev/libaums/usb/AndroidUsbCommunication.kt
@@ -60,7 +60,11 @@ internal abstract class AndroidUsbCommunication(
     private fun nativeReset() {
         Log.d(TAG, "Performing native reset")
         val result = resetUsbDeviceNative(deviceConnection!!.fileDescriptor)
-        Log.d(TAG, "ioctl returned $result; errno ${ErrNo.errno} ${ErrNo.errstr}; ignoring")
+
+        Log.d(TAG, "ioctl okay: $result; errno ${ErrNo.errno} ${ErrNo.errstr}")
+        if (!result)
+            Log.w(TAG, "ioctl failed! USB device will likely require new discovery and permissions")
+
         Thread.sleep(1000)
 
         Log.d(TAG, "Closing and reopening USB device")

--- a/libaums/src/main/java/com/github/mjdev/libaums/usb/AndroidUsbCommunication.kt
+++ b/libaums/src/main/java/com/github/mjdev/libaums/usb/AndroidUsbCommunication.kt
@@ -52,20 +52,29 @@ internal abstract class AndroidUsbCommunication(
         return deviceConnection!!.controlTransfer(requestType, request, value, index, buffer, length, TRANSFER_TIMEOUT)
     }
 
-    override fun resetRecovery() {
-        if (isNativeInited) {
-            Log.w(TAG, "Native reset")
-            val result = resetUsbDeviceNative(deviceConnection.fileDescriptor)
-            Log.w(TAG, "ioctl returned $result; errno ${ErrNo.errno} ${ErrNo.errstr}; ignoring")
-            Thread.sleep(2000)
-        } else {
-            bulkOnlyMassStorageReset()
-            Thread.sleep(2000)
-            clearFeatureHalt(inEndpoint)
-            Thread.sleep(2000)
-            clearFeatureHalt(outEndpoint)
-            Thread.sleep(2000)
-        }
+    override fun resetRecovery() = when {
+        isNativeInited -> nativeReset()
+        else           -> softReset()
+    }
+
+    private fun nativeReset() {
+        Log.w(TAG, "Native reset")
+        val result = resetUsbDeviceNative(deviceConnection!!.fileDescriptor)
+        Log.w(TAG, "ioctl returned $result; errno ${ErrNo.errno} ${ErrNo.errstr}; ignoring")
+        Thread.sleep(1000)
+
+        // Reopen UsbConnection
+        closeUsbConnection()
+        initUsbConnection()
+    }
+
+    private fun softReset() {
+        bulkOnlyMassStorageReset()
+        Thread.sleep(2000)
+        clearFeatureHalt(inEndpoint)
+        Thread.sleep(2000)
+        clearFeatureHalt(outEndpoint)
+        Thread.sleep(2000)
     }
 
     override fun bulkOnlyMassStorageReset() {

--- a/libaums/src/main/java/com/github/mjdev/libaums/usb/AndroidUsbCommunication.kt
+++ b/libaums/src/main/java/com/github/mjdev/libaums/usb/AndroidUsbCommunication.kt
@@ -11,8 +11,8 @@ internal abstract class AndroidUsbCommunication(
         private val usbManager: UsbManager,
         private val usbDevice: UsbDevice,
         private val usbInterface: UsbInterface,
-        private val outEndpoint: UsbEndpoint,
-        private val inEndpoint: UsbEndpoint
+        val outEndpoint: UsbEndpoint,
+        val inEndpoint: UsbEndpoint
 ) : UsbCommunication {
 
     private var isNativeInited: Boolean = false
@@ -73,7 +73,7 @@ internal abstract class AndroidUsbCommunication(
     }
 
     private fun softReset() {
-        Log.w(TAG, "Native helper library not available, can't perform proper device reset")
+        Log.w(TAG, "Native helper library not available, can't perform hard device reset")
         bulkOnlyMassStorageReset()
         Thread.sleep(2000)
         clearFeatureHalt(inEndpoint)
@@ -94,18 +94,12 @@ internal abstract class AndroidUsbCommunication(
     }
 
     override fun clearFeatureHalt(endpoint: UsbEndpoint) {
-        Log.w(TAG, "sending clear feature halt")
-        val bArr = ByteArray(2)
-        val address: Int = endpoint.address
-        // REQUEST_CLEAR_FEATURE = 1
-        // REQUEST_TYPE_CLEAR_FEATURE = 2
-        val transferred: Int = controlTransfer(2, 1, 0, address, bArr, 0)
-        if (transferred == -1) {
-            throw IOException("bulk only mass storage reset failed!")
+        Log.w(TAG, "Clearing halt on endpoint $endpoint (direction ${endpoint.direction})")
+        val result = clearHaltNative(deviceConnection!!.fileDescriptor, endpoint.address)
+        if (!result) {
+            Log.e(TAG, "Clear halt failed: errno ${ErrNo.errno} ${ErrNo.errstr}")
         }
     }
-
-    private external fun resetUsbDeviceNative(fd: Int): Boolean
 
     private fun closeUsbConnection() {
         if (deviceConnection == null)
@@ -128,4 +122,8 @@ internal abstract class AndroidUsbCommunication(
     companion object {
         private val TAG = AndroidUsbCommunication::class.java.simpleName
     }
+
+    private external fun resetUsbDeviceNative(fd: Int): Boolean
+    private external fun clearHaltNative(fd: Int, endpoint: Int): Boolean
+
 }

--- a/libaums/src/main/java/com/github/mjdev/libaums/usb/AndroidUsbCommunication.kt
+++ b/libaums/src/main/java/com/github/mjdev/libaums/usb/AndroidUsbCommunication.kt
@@ -3,6 +3,7 @@ package com.github.mjdev.libaums.usb
 import android.hardware.usb.UsbDeviceConnection
 import android.hardware.usb.UsbEndpoint
 import android.hardware.usb.UsbInterface
+import android.util.Log
 import com.github.mjdev.libaums.usb.UsbCommunication.Companion.TRANSFER_TIMEOUT
 import java.io.IOException
 
@@ -13,11 +14,16 @@ internal abstract class AndroidUsbCommunication(private val deviceConnection: Us
     }
 
     override fun resetRecovery() {
-        bulkOnlyMassStorageReset();
-        clearFeatureHalt(inEndpoint);
+        bulkOnlyMassStorageReset()
+        Thread.sleep(10000)
+        clearFeatureHalt(inEndpoint)
+        Thread.sleep(10000)
+        clearFeatureHalt(outEndpoint)
+        Thread.sleep(10000)
     }
 
     override fun bulkOnlyMassStorageReset() {
+        Log.w(TAG, "sending bulk only mass storage request")
         val bArr = ByteArray(2)
         // REQUEST_BULK_ONLY_MASS_STORAGE_RESET = 255
         // REQUEST_TYPE_BULK_ONLY_MASS_STORAGE_RESET = 33
@@ -28,6 +34,7 @@ internal abstract class AndroidUsbCommunication(private val deviceConnection: Us
     }
 
     override fun clearFeatureHalt(endpoint: UsbEndpoint) {
+        Log.w(TAG, "sending clear feature halt")
         val bArr = ByteArray(2)
         val address: Int = endpoint.address
         // REQUEST_CLEAR_FEATURE = 1
@@ -36,5 +43,9 @@ internal abstract class AndroidUsbCommunication(private val deviceConnection: Us
         if (transferred == -1) {
             throw IOException("bulk only mass storage reset failed!")
         }
+    }
+
+    companion object {
+        private val TAG = AndroidUsbCommunication::class.java.simpleName
     }
 }

--- a/libaums/src/main/java/com/github/mjdev/libaums/usb/AndroidUsbCommunication.kt
+++ b/libaums/src/main/java/com/github/mjdev/libaums/usb/AndroidUsbCommunication.kt
@@ -58,17 +58,18 @@ internal abstract class AndroidUsbCommunication(
     }
 
     private fun nativeReset() {
-        Log.w(TAG, "Native reset")
+        Log.d(TAG, "Performing native reset")
         val result = resetUsbDeviceNative(deviceConnection!!.fileDescriptor)
-        Log.w(TAG, "ioctl returned $result; errno ${ErrNo.errno} ${ErrNo.errstr}; ignoring")
+        Log.d(TAG, "ioctl returned $result; errno ${ErrNo.errno} ${ErrNo.errstr}; ignoring")
         Thread.sleep(1000)
 
-        // Reopen UsbConnection
+        Log.d(TAG, "Closing and reopening USB device")
         closeUsbConnection()
         initUsbConnection()
     }
 
     private fun softReset() {
+        Log.w(TAG, "Native helper library not available, can't perform proper device reset")
         bulkOnlyMassStorageReset()
         Thread.sleep(2000)
         clearFeatureHalt(inEndpoint)

--- a/libaums/src/main/java/com/github/mjdev/libaums/usb/AndroidUsbCommunication.kt
+++ b/libaums/src/main/java/com/github/mjdev/libaums/usb/AndroidUsbCommunication.kt
@@ -9,6 +9,20 @@ import java.io.IOException
 
 
 internal abstract class AndroidUsbCommunication(private val deviceConnection: UsbDeviceConnection, private val usbInterface: UsbInterface, private val outEndpoint: UsbEndpoint, private val inEndpoint: UsbEndpoint) : UsbCommunication {
+
+    private var isNativeInited: Boolean = false
+
+    init {
+        try {
+            System.loadLibrary("usb-lib")
+            isNativeInited = true
+        } catch (e: UnsatisfiedLinkError) {
+            isNativeInited = false
+            Log.e(TAG, "could not load usb-lib", e)
+        }
+
+    }
+
     override fun controlTransfer(requestType: Int, request: Int, value: Int, index: Int, buffer: ByteArray, length: Int): Int {
         return deviceConnection.controlTransfer(requestType, request, value, index, buffer, length, TRANSFER_TIMEOUT)
     }
@@ -19,6 +33,14 @@ internal abstract class AndroidUsbCommunication(private val deviceConnection: Us
         clearFeatureHalt(inEndpoint)
         Thread.sleep(10000)
         clearFeatureHalt(outEndpoint)
+        Thread.sleep(10000)
+
+        if (isNativeInited) {
+            Log.w(TAG, "Native reset")
+            if (!resetUsbDeviceNative(deviceConnection.fileDescriptor)) {
+                throw IOException("native reset unsuccessful!")
+            }
+        }
         Thread.sleep(10000)
     }
 
@@ -44,6 +66,8 @@ internal abstract class AndroidUsbCommunication(private val deviceConnection: Us
             throw IOException("bulk only mass storage reset failed!")
         }
     }
+
+    private external fun resetUsbDeviceNative(fd: Int): Boolean
 
     companion object {
         private val TAG = AndroidUsbCommunication::class.java.simpleName

--- a/libaums/src/main/java/com/github/mjdev/libaums/usb/AndroidUsbCommunication.kt
+++ b/libaums/src/main/java/com/github/mjdev/libaums/usb/AndroidUsbCommunication.kt
@@ -4,6 +4,7 @@ import android.hardware.usb.UsbDeviceConnection
 import android.hardware.usb.UsbEndpoint
 import android.hardware.usb.UsbInterface
 import android.util.Log
+import com.github.mjdev.libaums.ErrNo
 import com.github.mjdev.libaums.usb.UsbCommunication.Companion.TRANSFER_TIMEOUT
 import java.io.IOException
 
@@ -28,19 +29,19 @@ internal abstract class AndroidUsbCommunication(private val deviceConnection: Us
     }
 
     override fun resetRecovery() {
-        bulkOnlyMassStorageReset()
-        Thread.sleep(10000)
-        clearFeatureHalt(inEndpoint)
-        Thread.sleep(10000)
-        clearFeatureHalt(outEndpoint)
-        Thread.sleep(10000)
-
         if (isNativeInited) {
             Log.w(TAG, "Native reset")
             val result = resetUsbDeviceNative(deviceConnection.fileDescriptor)
-            Log.d(TAG, "ioctl returned $result")
+            Log.w(TAG, "ioctl returned $result; errno ${ErrNo.errno} ${ErrNo.errstr}; ignoring")
+            Thread.sleep(2000)
+        } else {
+            bulkOnlyMassStorageReset()
+            Thread.sleep(2000)
+            clearFeatureHalt(inEndpoint)
+            Thread.sleep(2000)
+            clearFeatureHalt(outEndpoint)
+            Thread.sleep(2000)
         }
-        Thread.sleep(10000)
     }
 
     override fun bulkOnlyMassStorageReset() {

--- a/libaums/src/main/java/com/github/mjdev/libaums/usb/AndroidUsbCommunication.kt
+++ b/libaums/src/main/java/com/github/mjdev/libaums/usb/AndroidUsbCommunication.kt
@@ -1,19 +1,44 @@
 package com.github.mjdev.libaums.usb
 
-import android.hardware.usb.UsbDeviceConnection
-import android.hardware.usb.UsbEndpoint
-import android.hardware.usb.UsbInterface
+import android.hardware.usb.*
 import android.util.Log
 import com.github.mjdev.libaums.ErrNo
 import com.github.mjdev.libaums.usb.UsbCommunication.Companion.TRANSFER_TIMEOUT
 import java.io.IOException
 
 
-internal abstract class AndroidUsbCommunication(private val deviceConnection: UsbDeviceConnection, private val usbInterface: UsbInterface, private val outEndpoint: UsbEndpoint, private val inEndpoint: UsbEndpoint) : UsbCommunication {
+internal abstract class AndroidUsbCommunication(
+        private val usbManager: UsbManager,
+        private val usbDevice: UsbDevice,
+        private val usbInterface: UsbInterface,
+        private val outEndpoint: UsbEndpoint,
+        private val inEndpoint: UsbEndpoint
+) : UsbCommunication {
 
     private var isNativeInited: Boolean = false
+    var deviceConnection: UsbDeviceConnection? = null
+    private var isClosed = false
 
     init {
+        initNativeLibrary()
+        initUsbConnection()
+    }
+
+    private fun initUsbConnection() {
+        if (isClosed)
+            return
+
+        Log.d(TAG, "setup device")
+        deviceConnection = usbManager.openDevice(usbDevice)
+                ?: throw IOException("deviceConnection is null!")
+
+        val claim = deviceConnection!!.claimInterface(usbInterface, true)
+        if (!claim) {
+            throw IOException("could not claim interface!")
+        }
+    }
+
+    private fun initNativeLibrary() {
         try {
             System.loadLibrary("usb-lib")
             isNativeInited = true
@@ -21,11 +46,10 @@ internal abstract class AndroidUsbCommunication(private val deviceConnection: Us
             isNativeInited = false
             Log.e(TAG, "could not load usb-lib", e)
         }
-
     }
 
     override fun controlTransfer(requestType: Int, request: Int, value: Int, index: Int, buffer: ByteArray, length: Int): Int {
-        return deviceConnection.controlTransfer(requestType, request, value, index, buffer, length, TRANSFER_TIMEOUT)
+        return deviceConnection!!.controlTransfer(requestType, request, value, index, buffer, length, TRANSFER_TIMEOUT)
     }
 
     override fun resetRecovery() {
@@ -68,6 +92,24 @@ internal abstract class AndroidUsbCommunication(private val deviceConnection: Us
     }
 
     private external fun resetUsbDeviceNative(fd: Int): Boolean
+
+    private fun closeUsbConnection() {
+        if (deviceConnection == null)
+            return
+
+        val release = deviceConnection!!.releaseInterface(usbInterface)
+        if (!release) {
+            Log.e(TAG, "could not release interface!")
+        }
+
+        deviceConnection!!.close()
+    }
+
+    override fun close() {
+        Log.d(TAG, "close device")
+        closeUsbConnection()
+        isClosed = true
+    }
 
     companion object {
         private val TAG = AndroidUsbCommunication::class.java.simpleName

--- a/libaums/src/main/java/com/github/mjdev/libaums/usb/AndroidUsbCommunication.kt
+++ b/libaums/src/main/java/com/github/mjdev/libaums/usb/AndroidUsbCommunication.kt
@@ -37,9 +37,8 @@ internal abstract class AndroidUsbCommunication(private val deviceConnection: Us
 
         if (isNativeInited) {
             Log.w(TAG, "Native reset")
-            if (!resetUsbDeviceNative(deviceConnection.fileDescriptor)) {
-                throw IOException("native reset unsuccessful!")
-            }
+            val result = resetUsbDeviceNative(deviceConnection.fileDescriptor)
+            Log.d(TAG, "ioctl returned $result")
         }
         Thread.sleep(10000)
     }

--- a/libaums/src/main/java/com/github/mjdev/libaums/usb/AndroidUsbCommunication.kt
+++ b/libaums/src/main/java/com/github/mjdev/libaums/usb/AndroidUsbCommunication.kt
@@ -59,22 +59,21 @@ internal abstract class AndroidUsbCommunication(
 
     private fun nativeReset() {
         Log.d(TAG, "Performing native reset")
+        val result = resetUsbDeviceNative(deviceConnection!!.fileDescriptor)
 
-        if (!deviceConnection!!.releaseInterface(usbInterface)) {
-            Log.w(TAG, "Failed to release interface, errno: ${ErrNo.errno} ${ErrNo.errstr}")
-        }
+        Log.d(TAG, "ioctl okay: $result; errno ${ErrNo.errno} ${ErrNo.errstr}")
+        if (!result)
+            Log.w(TAG, "ioctl failed! USB device will likely require new discovery and permissions")
 
-        if (!resetUsbDeviceNative(deviceConnection!!.fileDescriptor)) {
-            Log.w(TAG, "ioctl failed! errno ${ErrNo.errno} ${ErrNo.errstr}")
-            Log.w(TAG, "USB device will likely require new discovery and permissions")
-        }
+        Thread.sleep(1000)
 
-        if (!deviceConnection!!.claimInterface(usbInterface, true)) {
-            throw IOException("Could not claim interface, errno: ${ErrNo.errno} ${ErrNo.errstr}")
-        }
+        Log.d(TAG, "Closing and reopening USB device")
+        closeUsbConnection()
+        initUsbConnection()
     }
 
     private fun softReset() {
+        Log.w(TAG, "Native helper library not available, can't perform hard device reset")
         bulkOnlyMassStorageReset()
         Thread.sleep(2000)
         clearFeatureHalt(inEndpoint)

--- a/libaums/src/main/java/com/github/mjdev/libaums/usb/HoneyCombMr1Communication.kt
+++ b/libaums/src/main/java/com/github/mjdev/libaums/usb/HoneyCombMr1Communication.kt
@@ -17,8 +17,8 @@ internal class HoneyCombMr1Communication(
         usbManager: UsbManager,
         usbDevice: UsbDevice,
         usbInterface: UsbInterface,
-        private val outEndpoint: UsbEndpoint,
-        private val inEndpoint: UsbEndpoint
+        outEndpoint: UsbEndpoint,
+        inEndpoint: UsbEndpoint
 ) : AndroidUsbCommunication(usbManager, usbDevice, usbInterface, outEndpoint, inEndpoint) {
 
     @Throws(IOException::class)

--- a/libaums/src/main/java/com/github/mjdev/libaums/usb/HoneyCombMr1Communication.kt
+++ b/libaums/src/main/java/com/github/mjdev/libaums/usb/HoneyCombMr1Communication.kt
@@ -2,6 +2,7 @@ package com.github.mjdev.libaums.usb
 
 import android.hardware.usb.UsbDeviceConnection
 import android.hardware.usb.UsbEndpoint
+import android.hardware.usb.UsbInterface
 
 import java.io.IOException
 import java.nio.ByteBuffer
@@ -14,7 +15,7 @@ import java.nio.ByteBuffer
  *
  * @author mjahnen
  */
-internal class HoneyCombMr1Communication(private val deviceConnection: UsbDeviceConnection, private val outEndpoint: UsbEndpoint, private val inEndpoint: UsbEndpoint) : UsbCommunication {
+internal class HoneyCombMr1Communication(private val deviceConnection: UsbDeviceConnection, private val usbInterface: UsbInterface, private val outEndpoint: UsbEndpoint, private val inEndpoint: UsbEndpoint) : AndroidUsbCommunication(deviceConnection, usbInterface, outEndpoint, inEndpoint) {
 
     @Throws(IOException::class)
     override fun bulkOutTransfer(src: ByteBuffer): Int {

--- a/libaums/src/main/java/com/github/mjdev/libaums/usb/HoneyCombMr1Communication.kt
+++ b/libaums/src/main/java/com/github/mjdev/libaums/usb/HoneyCombMr1Communication.kt
@@ -1,8 +1,6 @@
 package com.github.mjdev.libaums.usb
 
-import android.hardware.usb.UsbDeviceConnection
-import android.hardware.usb.UsbEndpoint
-import android.hardware.usb.UsbInterface
+import android.hardware.usb.*
 
 import java.io.IOException
 import java.nio.ByteBuffer
@@ -15,14 +13,20 @@ import java.nio.ByteBuffer
  *
  * @author mjahnen
  */
-internal class HoneyCombMr1Communication(private val deviceConnection: UsbDeviceConnection, private val usbInterface: UsbInterface, private val outEndpoint: UsbEndpoint, private val inEndpoint: UsbEndpoint) : AndroidUsbCommunication(deviceConnection, usbInterface, outEndpoint, inEndpoint) {
+internal class HoneyCombMr1Communication(
+        usbManager: UsbManager,
+        usbDevice: UsbDevice,
+        usbInterface: UsbInterface,
+        private val outEndpoint: UsbEndpoint,
+        private val inEndpoint: UsbEndpoint
+) : AndroidUsbCommunication(usbManager, usbDevice, usbInterface, outEndpoint, inEndpoint) {
 
     @Throws(IOException::class)
     override fun bulkOutTransfer(src: ByteBuffer): Int {
         val offset = src.position()
 
         if (offset == 0) {
-            val result = deviceConnection.bulkTransfer(outEndpoint,
+            val result = deviceConnection!!.bulkTransfer(outEndpoint,
                     src.array(), src.remaining(), UsbCommunication.TRANSFER_TIMEOUT)
 
             if (result == -1) {
@@ -35,7 +39,7 @@ internal class HoneyCombMr1Communication(private val deviceConnection: UsbDevice
 
         val tmpBuffer = ByteArray(src.remaining())
         System.arraycopy(src.array(), offset, tmpBuffer, 0, src.remaining())
-        val result = deviceConnection.bulkTransfer(outEndpoint,
+        val result = deviceConnection!!.bulkTransfer(outEndpoint,
                 tmpBuffer, src.remaining(), UsbCommunication.TRANSFER_TIMEOUT)
 
         if (result == -1) {
@@ -51,7 +55,7 @@ internal class HoneyCombMr1Communication(private val deviceConnection: UsbDevice
         val offset = dest.position()
 
         if (offset == 0) {
-            val result = deviceConnection.bulkTransfer(inEndpoint,
+            val result = deviceConnection!!.bulkTransfer(inEndpoint,
                     dest.array(), dest.remaining(), UsbCommunication.TRANSFER_TIMEOUT)
 
             if (result == -1) {
@@ -64,7 +68,7 @@ internal class HoneyCombMr1Communication(private val deviceConnection: UsbDevice
         }
 
         val tmpBuffer = ByteArray(dest.remaining())
-        val result = deviceConnection.bulkTransfer(inEndpoint, tmpBuffer, dest.remaining(), UsbCommunication.TRANSFER_TIMEOUT)
+        val result = deviceConnection!!.bulkTransfer(inEndpoint, tmpBuffer, dest.remaining(), UsbCommunication.TRANSFER_TIMEOUT)
 
         if (result == -1) {
             throw IOException("Could not read from device, result == -1")

--- a/libaums/src/main/java/com/github/mjdev/libaums/usb/JellyBeanMr2Communication.kt
+++ b/libaums/src/main/java/com/github/mjdev/libaums/usb/JellyBeanMr2Communication.kt
@@ -5,14 +5,9 @@ package com.github.mjdev.libaums.usb
  */
 
 import android.annotation.TargetApi
-import android.hardware.usb.UsbDeviceConnection
-import android.hardware.usb.UsbEndpoint
-import android.hardware.usb.UsbInterface
+import android.hardware.usb.*
 import android.os.Build
-import android.system.ErrnoException
-
 import com.github.mjdev.libaums.ErrNo
-
 import java.io.IOException
 import java.nio.ByteBuffer
 
@@ -24,11 +19,17 @@ import java.nio.ByteBuffer
  * @author mjahnen
  */
 @TargetApi(Build.VERSION_CODES.JELLY_BEAN_MR2)
-internal class JellyBeanMr2Communication(private val deviceConnection: UsbDeviceConnection, private val usbInterface: UsbInterface, private val outEndpoint: UsbEndpoint, private val inEndpoint: UsbEndpoint) : AndroidUsbCommunication(deviceConnection, usbInterface, outEndpoint, inEndpoint) {
+internal class JellyBeanMr2Communication(
+        usbManager: UsbManager,
+        usbDevice: UsbDevice,
+        usbInterface: UsbInterface,
+        private val outEndpoint: UsbEndpoint,
+        private val inEndpoint: UsbEndpoint
+) : AndroidUsbCommunication(usbManager, usbDevice, usbInterface, outEndpoint, inEndpoint) {
 
     @Throws(IOException::class)
     override fun bulkOutTransfer(src: ByteBuffer): Int {
-        val result = deviceConnection.bulkTransfer(outEndpoint,
+        val result = deviceConnection!!.bulkTransfer(outEndpoint,
                 src.array(), src.position(), src.remaining(), UsbCommunication.TRANSFER_TIMEOUT)
 
         if (result == -1) {
@@ -41,7 +42,7 @@ internal class JellyBeanMr2Communication(private val deviceConnection: UsbDevice
 
     @Throws(IOException::class)
     override fun bulkInTransfer(dest: ByteBuffer): Int {
-        val result = deviceConnection.bulkTransfer(inEndpoint,
+        val result = deviceConnection!!.bulkTransfer(inEndpoint,
                 dest.array(), dest.position(), dest.remaining(), UsbCommunication.TRANSFER_TIMEOUT)
 
         if (result == -1) {

--- a/libaums/src/main/java/com/github/mjdev/libaums/usb/JellyBeanMr2Communication.kt
+++ b/libaums/src/main/java/com/github/mjdev/libaums/usb/JellyBeanMr2Communication.kt
@@ -23,8 +23,8 @@ internal class JellyBeanMr2Communication(
         usbManager: UsbManager,
         usbDevice: UsbDevice,
         usbInterface: UsbInterface,
-        private val outEndpoint: UsbEndpoint,
-        private val inEndpoint: UsbEndpoint
+        outEndpoint: UsbEndpoint,
+        inEndpoint: UsbEndpoint
 ) : AndroidUsbCommunication(usbManager, usbDevice, usbInterface, outEndpoint, inEndpoint) {
 
     @Throws(IOException::class)

--- a/libaums/src/main/java/com/github/mjdev/libaums/usb/JellyBeanMr2Communication.kt
+++ b/libaums/src/main/java/com/github/mjdev/libaums/usb/JellyBeanMr2Communication.kt
@@ -7,6 +7,7 @@ package com.github.mjdev.libaums.usb
 import android.annotation.TargetApi
 import android.hardware.usb.UsbDeviceConnection
 import android.hardware.usb.UsbEndpoint
+import android.hardware.usb.UsbInterface
 import android.os.Build
 import android.system.ErrnoException
 
@@ -23,7 +24,7 @@ import java.nio.ByteBuffer
  * @author mjahnen
  */
 @TargetApi(Build.VERSION_CODES.JELLY_BEAN_MR2)
-internal class JellyBeanMr2Communication(private val deviceConnection: UsbDeviceConnection, private val outEndpoint: UsbEndpoint, private val inEndpoint: UsbEndpoint) : UsbCommunication {
+internal class JellyBeanMr2Communication(private val deviceConnection: UsbDeviceConnection, private val usbInterface: UsbInterface, private val outEndpoint: UsbEndpoint, private val inEndpoint: UsbEndpoint) : AndroidUsbCommunication(deviceConnection, usbInterface, outEndpoint, inEndpoint) {
 
     @Throws(IOException::class)
     override fun bulkOutTransfer(src: ByteBuffer): Int {

--- a/libaums/src/main/java/com/github/mjdev/libaums/usb/UsbCommunication.kt
+++ b/libaums/src/main/java/com/github/mjdev/libaums/usb/UsbCommunication.kt
@@ -54,13 +54,36 @@ interface UsbCommunication : Closeable {
     @Throws(IOException::class)
     fun bulkInTransfer(dest: ByteBuffer): Int
 
+    /**
+     * Performs a control transaction on endpoint zero for this device.
+     *
+     * @param requestType request type for this transaction
+     * @param request request ID for this transaction
+     * @param value value field for this transaction
+     * @param index index field for this transaction
+     * @param buffer buffer for data portion of transaction, or null if no data needs to be sent or received
+     * @param length the length of the data to send or receive
+     */
     @Throws(IOException::class)
     fun controlTransfer(requestType: Int, request: Int, value: Int, index: Int, buffer: ByteArray, length: Int): Int
 
+    /**
+     * Performs the recovery reset procedure in case the communication is stalled
+     */
     @Throws(IOException::class)
     fun resetRecovery()
+
+    /**
+     * Attempts a bulk-only mass storage reset control transfer operation
+     */
     @Throws(IOException::class)
     fun bulkOnlyMassStorageReset()
+
+    /**
+     * Attempts to clear the HALT feature from the specified USB endpoint
+     *
+     * @param endpoint the USB endpoint to clear the HALT feature from
+     */
     @Throws(IOException::class)
     fun clearFeatureHalt(endpoint: UsbEndpoint)
 

--- a/libaums/src/main/java/com/github/mjdev/libaums/usb/UsbCommunication.kt
+++ b/libaums/src/main/java/com/github/mjdev/libaums/usb/UsbCommunication.kt
@@ -31,6 +31,8 @@ import java.nio.ByteBuffer
  * @author mjahnen
  */
 interface UsbCommunication : Closeable {
+    val inEndpoint: UsbEndpoint
+    val outEndpoint: UsbEndpoint
 
     /**
      * Performs a bulk out transfer beginning at the offset specified in the

--- a/libaums/src/main/java/com/github/mjdev/libaums/usb/UsbCommunication.kt
+++ b/libaums/src/main/java/com/github/mjdev/libaums/usb/UsbCommunication.kt
@@ -18,6 +18,7 @@
 package com.github.mjdev.libaums.usb
 
 import android.hardware.usb.UsbEndpoint
+import java.io.Closeable
 import java.io.IOException
 import java.nio.ByteBuffer
 
@@ -29,7 +30,7 @@ import java.nio.ByteBuffer
  *
  * @author mjahnen
  */
-interface UsbCommunication {
+interface UsbCommunication : Closeable {
 
     /**
      * Performs a bulk out transfer beginning at the offset specified in the

--- a/libaums/src/main/java/com/github/mjdev/libaums/usb/UsbCommunication.kt
+++ b/libaums/src/main/java/com/github/mjdev/libaums/usb/UsbCommunication.kt
@@ -54,6 +54,7 @@ interface UsbCommunication : Closeable {
     @Throws(IOException::class)
     fun bulkInTransfer(dest: ByteBuffer): Int
 
+    @Throws(IOException::class)
     fun controlTransfer(requestType: Int, request: Int, value: Int, index: Int, buffer: ByteArray, length: Int): Int
 
     @Throws(IOException::class)

--- a/libaums/src/main/java/com/github/mjdev/libaums/usb/UsbCommunication.kt
+++ b/libaums/src/main/java/com/github/mjdev/libaums/usb/UsbCommunication.kt
@@ -17,6 +17,7 @@
 
 package com.github.mjdev.libaums.usb
 
+import android.hardware.usb.UsbEndpoint
 import java.io.IOException
 import java.nio.ByteBuffer
 
@@ -51,6 +52,15 @@ interface UsbCommunication {
      */
     @Throws(IOException::class)
     fun bulkInTransfer(dest: ByteBuffer): Int
+
+    fun controlTransfer(requestType: Int, request: Int, value: Int, index: Int, buffer: ByteArray, length: Int): Int
+
+    @Throws(IOException::class)
+    fun resetRecovery()
+    @Throws(IOException::class)
+    fun bulkOnlyMassStorageReset()
+    @Throws(IOException::class)
+    fun clearFeatureHalt(endpoint: UsbEndpoint)
 
     companion object {
         const val TRANSFER_TIMEOUT = 5000

--- a/libaums/src/main/java/com/github/mjdev/libaums/usb/UsbCommunicationFactory.kt
+++ b/libaums/src/main/java/com/github/mjdev/libaums/usb/UsbCommunicationFactory.kt
@@ -1,8 +1,9 @@
 package com.github.mjdev.libaums.usb
 
-import android.hardware.usb.UsbDeviceConnection
+import android.hardware.usb.UsbDevice
 import android.hardware.usb.UsbEndpoint
 import android.hardware.usb.UsbInterface
+import android.hardware.usb.UsbManager
 import android.os.Build
 import android.util.Log
 
@@ -22,16 +23,22 @@ object UsbCommunicationFactory {
         DEVICE_CONNECTION_SYNC
     }
 
-    fun createUsbCommunication(deviceConnection: UsbDeviceConnection, usbInterface: UsbInterface, outEndpoint: UsbEndpoint, inEndpoint: UsbEndpoint): UsbCommunication {
+    fun createUsbCommunication(
+            usbManager: UsbManager,
+            usbDevice: UsbDevice,
+            usbInterface: UsbInterface,
+            outEndpoint: UsbEndpoint,
+            inEndpoint: UsbEndpoint
+    ): UsbCommunication {
         return if (underlyingUsbCommunication == UnderlyingUsbCommunication.DEVICE_CONNECTION_SYNC) {
             if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR2) {
-                JellyBeanMr2Communication(deviceConnection, usbInterface, outEndpoint, inEndpoint)
+                JellyBeanMr2Communication(usbManager, usbDevice, usbInterface, outEndpoint, inEndpoint)
             } else {
                 Log.i(TAG, "using workaround usb communication")
-                HoneyCombMr1Communication(deviceConnection, usbInterface, outEndpoint, inEndpoint)
+                HoneyCombMr1Communication(usbManager, usbDevice, usbInterface, outEndpoint, inEndpoint)
             }
         } else {
-            UsbRequestCommunication(deviceConnection, usbInterface, outEndpoint, inEndpoint)
+            UsbRequestCommunication(usbManager, usbDevice, usbInterface, outEndpoint, inEndpoint)
         }
     }
 }

--- a/libaums/src/main/java/com/github/mjdev/libaums/usb/UsbCommunicationFactory.kt
+++ b/libaums/src/main/java/com/github/mjdev/libaums/usb/UsbCommunicationFactory.kt
@@ -2,6 +2,7 @@ package com.github.mjdev.libaums.usb
 
 import android.hardware.usb.UsbDeviceConnection
 import android.hardware.usb.UsbEndpoint
+import android.hardware.usb.UsbInterface
 import android.os.Build
 import android.util.Log
 
@@ -21,16 +22,16 @@ object UsbCommunicationFactory {
         DEVICE_CONNECTION_SYNC
     }
 
-    fun createUsbCommunication(deviceConnection: UsbDeviceConnection, outEndpoint: UsbEndpoint, inEndpoint: UsbEndpoint): UsbCommunication {
+    fun createUsbCommunication(deviceConnection: UsbDeviceConnection, usbInterface: UsbInterface, outEndpoint: UsbEndpoint, inEndpoint: UsbEndpoint): UsbCommunication {
         return if (underlyingUsbCommunication == UnderlyingUsbCommunication.DEVICE_CONNECTION_SYNC) {
             if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR2) {
-                JellyBeanMr2Communication(deviceConnection, outEndpoint, inEndpoint)
+                JellyBeanMr2Communication(deviceConnection, usbInterface, outEndpoint, inEndpoint)
             } else {
                 Log.i(TAG, "using workaround usb communication")
-                HoneyCombMr1Communication(deviceConnection, outEndpoint, inEndpoint)
+                HoneyCombMr1Communication(deviceConnection, usbInterface, outEndpoint, inEndpoint)
             }
         } else {
-            UsbRequestCommunication(deviceConnection, outEndpoint, inEndpoint)
+            UsbRequestCommunication(deviceConnection, usbInterface, outEndpoint, inEndpoint)
         }
     }
 }

--- a/libaums/src/main/java/com/github/mjdev/libaums/usb/UsbRequestCommunication.kt
+++ b/libaums/src/main/java/com/github/mjdev/libaums/usb/UsbRequestCommunication.kt
@@ -13,8 +13,8 @@ internal class UsbRequestCommunication(
         usbManager: UsbManager,
         usbDevice: UsbDevice,
         usbInterface: UsbInterface,
-        private val outEndpoint: UsbEndpoint,
-        private val inEndpoint: UsbEndpoint
+        outEndpoint: UsbEndpoint,
+        inEndpoint: UsbEndpoint
 ) : AndroidUsbCommunication(usbManager, usbDevice, usbInterface, outEndpoint, inEndpoint) {
 
     private val outRequest = UsbRequest().apply { initialize(deviceConnection, outEndpoint) }

--- a/libaums/src/main/java/com/github/mjdev/libaums/usb/UsbRequestCommunication.kt
+++ b/libaums/src/main/java/com/github/mjdev/libaums/usb/UsbRequestCommunication.kt
@@ -2,6 +2,7 @@ package com.github.mjdev.libaums.usb
 
 import android.hardware.usb.UsbDeviceConnection
 import android.hardware.usb.UsbEndpoint
+import android.hardware.usb.UsbInterface
 import android.hardware.usb.UsbRequest
 
 import java.io.IOException
@@ -11,7 +12,7 @@ import java.nio.ByteBuffer
  * Created by magnusja on 21/12/16.
  */
 
-internal class UsbRequestCommunication(private val deviceConnection: UsbDeviceConnection, outEndpoint: UsbEndpoint, inEndpoint: UsbEndpoint) : UsbCommunication {
+internal class UsbRequestCommunication(private val deviceConnection: UsbDeviceConnection, private val usbInterface: UsbInterface, outEndpoint: UsbEndpoint, inEndpoint: UsbEndpoint) : AndroidUsbCommunication(deviceConnection, usbInterface, outEndpoint, inEndpoint) {
     private val outRequest = UsbRequest().apply { initialize(deviceConnection, outEndpoint) }
     private val inRequest = UsbRequest().apply { initialize(deviceConnection, inEndpoint) }
     private val workaroundBuffer = ByteBuffer.allocate(1024 * 32 * 4)

--- a/libaums/src/main/java/com/github/mjdev/libaums/usb/UsbRequestCommunication.kt
+++ b/libaums/src/main/java/com/github/mjdev/libaums/usb/UsbRequestCommunication.kt
@@ -1,9 +1,6 @@
 package com.github.mjdev.libaums.usb
 
-import android.hardware.usb.UsbDeviceConnection
-import android.hardware.usb.UsbEndpoint
-import android.hardware.usb.UsbInterface
-import android.hardware.usb.UsbRequest
+import android.hardware.usb.*
 
 import java.io.IOException
 import java.nio.ByteBuffer
@@ -12,7 +9,14 @@ import java.nio.ByteBuffer
  * Created by magnusja on 21/12/16.
  */
 
-internal class UsbRequestCommunication(private val deviceConnection: UsbDeviceConnection, private val usbInterface: UsbInterface, outEndpoint: UsbEndpoint, inEndpoint: UsbEndpoint) : AndroidUsbCommunication(deviceConnection, usbInterface, outEndpoint, inEndpoint) {
+internal class UsbRequestCommunication(
+        usbManager: UsbManager,
+        usbDevice: UsbDevice,
+        usbInterface: UsbInterface,
+        private val outEndpoint: UsbEndpoint,
+        private val inEndpoint: UsbEndpoint
+) : AndroidUsbCommunication(usbManager, usbDevice, usbInterface, outEndpoint, inEndpoint) {
+
     private val outRequest = UsbRequest().apply { initialize(deviceConnection, outEndpoint) }
     private val inRequest = UsbRequest().apply { initialize(deviceConnection, inEndpoint) }
     private val workaroundBuffer = ByteBuffer.allocate(1024 * 32 * 4)
@@ -32,7 +36,7 @@ internal class UsbRequestCommunication(private val deviceConnection: UsbDeviceCo
             throw IOException("Error queueing request.")
         }
 
-        val request = deviceConnection.requestWait()
+        val request = deviceConnection!!.requestWait()
         if (request === outRequest) {
             src.position(oldPosition + workaroundBuffer.position())
             return workaroundBuffer.position()
@@ -54,7 +58,7 @@ internal class UsbRequestCommunication(private val deviceConnection: UsbDeviceCo
             throw IOException("Error queueing request.")
         }
 
-        val request = deviceConnection.requestWait()
+        val request = deviceConnection!!.requestWait()
         if (request === inRequest) {
             workaroundBuffer.flip()
             dest.put(workaroundBuffer)

